fem/bilininteg.cpp

```diff
@@ -553,17 +553,19 @@ void DiffusionIntegrator::AssembleElementMatrix
    double w;
 
 #ifdef MFEM_THREAD_SAFE
-   DenseMatrix dshape(nd,dim), dshapedxt(nd,spaceDim), invdfdx(dim,spaceDim);
+   DenseMatrix dshape(nd, dim), dshapedxt(nd, spaceDim);
+   DenseMatrix invdfdx(spaceDim, spaceDim);
    Vector D(VQ ? VQ->GetVDim() : 0);
 #else
-   dshape.SetSize(nd,dim);
-   dshapedxt.SetSize(nd,spaceDim);
-   invdfdx.SetSize(dim,spaceDim);
+   dshape.SetSize(nd, dim);
+   dshapedxt.SetSize(nd, spaceDim);
+   invdfdx.SetSize(spaceDim, spaceDim);
    D.SetSize(VQ ? VQ->GetVDim() : 0);
 #endif
    elmat.SetSize(nd);
 
    const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el, el);
+   DenseMatrix dshapedxt_m(nd, spaceDim);
```

Comment:
Since this is needed only inside the `if (MQ)` statement, it should be moved there

Reply:
Done

Reply:
Moving this definition inside the `if (MQ)` statement (which is inside the quadrature-point loop) will cause the constructor and destructor to be called for every quadrature point. This not optimal either. Maybe a better option is to always define `dshapedxt_m` but keep it empty:
```c++
   DenseMatrix dshapedxt_m;
```
And then do the allocation only conditionally, when needed:
```c++
   if (MQ) { dshapedxt_m.SetSize(nd, spaceDim); }
```

By the way, Mark, you said "Done" but I don't see the change in the branch -- maybe you forgot to push. However, I'm not really sure we want that change, considering the above point.

Reply:
good point!

Reply:
Yes, we're not moving the DenseMatrix inside the quadrature loop.

I pushed the necessary changes yesterday but it's not clear where they ended up.  I'm sure this is connected to the confusion @sshiraiwa is referring to below.